Flaky-test: V1_ProducerConsumerTest.testConcurrentConsumerReceiveWhileReconnect

org.apache.pulsar.client.api.v1.V1_ProducerConsumerTest is flaky. The testConcurrentConsumerReceiveWhileReconnect test method fails sporadically.

```
java.lang.AssertionError: expected [10] but found [0]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.client.api.v1.V1_ProducerConsumerTest.testConcurrentConsumerReceiveWhileReconnect(V1_ProducerConsumerTest.java:597)
```
Usage tip: To enable automatic navigation to failure message, open the links with CTRL/CMD-click.
[example failure 2021-02-01T11:51:46.1541792Z](https://github.com/apache/pulsar/runs/1805502919?check_suite_focus=true#step:9:1422)
[example failure 2021-02-01T11:31:54.7836221Z](https://github.com/apache/pulsar/runs/1805502919?check_suite_focus=true#step:9:543)
[example failure 2021-02-01T11:31:54.7720717Z](https://github.com/apache/pulsar/runs/1805502919?check_suite_focus=true#step:9:521)
[example failure 2021-02-01T10:16:07.0278884Z](https://github.com/apache/pulsar/runs/1804927633?check_suite_focus=true#step:9:1411)
[example failure 2021-02-01T06:23:48.3144807Z](https://github.com/apache/pulsar/runs/1804234652?check_suite_focus=true#step:9:521)
[example failure 2021-02-01T06:08:50.1003701Z](https://github.com/apache/pulsar/runs/1804227520?check_suite_focus=true#step:9:694)
[example failure 2021-02-01T02:50:56.7566310Z](https://github.com/apache/pulsar/runs/1803462376?check_suite_focus=true#step:9:1453)
[example failure 2021-02-01T02:41:01.3188597Z](https://github.com/apache/pulsar/runs/1803462376?check_suite_focus=true#step:9:995)
[example failure 2021-02-01T02:41:01.3074356Z](https://github.com/apache/pulsar/runs/1803462376?check_suite_focus=true#step:9:973)
[example failure 2021-01-31T08:33:49.7334910Z](https://github.com/apache/pulsar/runs/1800366490?check_suite_focus=true#step:9:1510)
[example failure 2021-01-31T08:25:13.9153525Z](https://github.com/apache/pulsar/runs/1800366490?check_suite_focus=true#step:9:1023)
[example failure 2021-01-30T16:32:32.2452165Z](https://github.com/apache/pulsar/runs/1797846048?check_suite_focus=true#step:9:1620)
[example failure 2021-01-30T16:21:39.3222001Z](https://github.com/apache/pulsar/runs/1797846048?check_suite_focus=true#step:9:1143)
[example failure 2021-01-30T16:11:15.7921611Z](https://github.com/apache/pulsar/runs/1797846048?check_suite_focus=true#step:9:721)
[example failure 2021-01-30T16:11:15.7831995Z](https://github.com/apache/pulsar/runs/1797846048?check_suite_focus=true#step:9:699)
[example failure 2021-01-30T08:16:58.3447205Z](https://github.com/apache/pulsar/runs/1796622804?check_suite_focus=true#step:8:549)
[example failure 2021-01-28T18:28:29.3671984Z](https://github.com/apache/pulsar/runs/1784534632?check_suite_focus=true#step:8:1436)
[example failure 2021-01-28T18:17:28.6125348Z](https://github.com/apache/pulsar/runs/1784534632?check_suite_focus=true#step:8:959)


<details>
<summary>Full exception stacktrace</summary>
<code><pre>
java.lang.AssertionError: expected [10] but found [0]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.client.api.v1.V1_ProducerConsumerTest.testConcurrentConsumerReceiveWhileReconnect(V1_ProducerConsumerTest.java:597)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

</pre></code>
</details>

```
java.lang.AssertionError: expected [10] but found [0]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.client.api.v1.V1_ProducerConsumerTest.testConcurrentConsumerReceiveWhileReconnect(V1_ProducerConsumerTest.java:545)
```
Usage tip: To enable automatic navigation to failure message, open the links with CTRL/CMD-click.
[example failure 2021-02-01T11:41:38.8272145Z](https://github.com/apache/pulsar/runs/1805502919?check_suite_focus=true#step:9:983)
[example failure 2021-02-01T09:54:32.3281809Z](https://github.com/apache/pulsar/runs/1804927633?check_suite_focus=true#step:9:521)
[example failure 2021-02-01T06:42:31.7239189Z](https://github.com/apache/pulsar/runs/1804234652?check_suite_focus=true#step:9:1584)
[example failure 2021-02-01T06:32:59.9627254Z](https://github.com/apache/pulsar/runs/1804234652?check_suite_focus=true#step:9:961)
[example failure 2021-02-01T06:29:39.4451179Z](https://github.com/apache/pulsar/runs/1804227520?check_suite_focus=true#step:9:1594)
[example failure 2021-02-01T06:19:09.3453561Z](https://github.com/apache/pulsar/runs/1804227520?check_suite_focus=true#step:9:1133)
[example failure 2021-01-31T08:33:49.7216494Z](https://github.com/apache/pulsar/runs/1800366490?check_suite_focus=true#step:9:1488)
[example failure 2021-01-31T08:25:13.9101460Z](https://github.com/apache/pulsar/runs/1800366490?check_suite_focus=true#step:9:1001)
[example failure 2021-01-31T08:16:30.6367524Z](https://github.com/apache/pulsar/runs/1800366490?check_suite_focus=true#step:9:521)


<details>
<summary>Full exception stacktrace</summary>
<code><pre>
java.lang.AssertionError: expected [10] but found [0]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.client.api.v1.V1_ProducerConsumerTest.testConcurrentConsumerReceiveWhileReconnect(V1_ProducerConsumerTest.java:545)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

</pre></code>
</details>

--- a/pulsar-broker/src/test/java/org/apache/pulsar/client/api/v1/V1_ProducerConsumerTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/client/api/v1/V1_ProducerConsumerTest.java
@@ -74,6 +74,7 @@ import org.apache.pulsar.client.impl.TypedMessageBuilderImpl;
 import org.apache.pulsar.common.protocol.Commands;
 import org.apache.pulsar.common.naming.TopicName;
 import org.apache.pulsar.common.util.FutureUtil;
+import org.awaitility.Awaitility;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.testng.Assert;
@@ -538,11 +539,11 @@ public class V1_ProducerConsumerTest extends V1_ProducerConsumerBase {
             String message = "my-message-" + i;
             producer.send(message.getBytes());
         }
-        Thread.sleep(500);
 
         ConsumerImpl<byte[]> consumerImpl = (ConsumerImpl<byte[]>) consumer;
         // The available permits should be 10 and num messages in the queue should be 90
-        Assert.assertEquals(consumerImpl.getAvailablePermits(), numConsumersThreads);
+        Awaitility.await().untilAsserted(() ->
+                Assert.assertEquals(consumerImpl.getAvailablePermits(), numConsumersThreads));
         Assert.assertEquals(consumerImpl.numMessagesInQueue(), recvQueueSize - numConsumersThreads);
 
         barrier.reset();
@@ -557,10 +558,10 @@ public class V1_ProducerConsumerTest extends V1_ProducerConsumerBase {
             });
         }
         barrier.await();
-        Thread.sleep(100);
 
         // The available permits should be 20 and num messages in the queue should be 80
-        Assert.assertEquals(consumerImpl.getAvailablePermits(), numConsumersThreads * 2);
+        Awaitility.await().untilAsserted(() ->
+                Assert.assertEquals(consumerImpl.getAvailablePermits(), numConsumersThreads * 2));
         Assert.assertEquals(consumerImpl.numMessagesInQueue(), recvQueueSize - (numConsumersThreads * 2));
 
         // clear the queue
@@ -591,10 +592,10 @@ public class V1_ProducerConsumerTest extends V1_ProducerConsumerBase {
         Thread.sleep(100);
 
         restartBroker();
-        Thread.sleep(2000);
 
         // The available permits should be 10 and num messages in the queue should be 90
-        Assert.assertEquals(consumerImpl.getAvailablePermits(), numConsumersThreads);
+        Awaitility.await().untilAsserted(() ->
+                Assert.assertEquals(consumerImpl.getAvailablePermits(), numConsumersThreads));
         Assert.assertEquals(consumerImpl.numMessagesInQueue(), recvQueueSize - numConsumersThreads);
         consumer.close();
 

